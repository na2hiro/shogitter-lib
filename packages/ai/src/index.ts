export * from "./playerRegistry.js";
export { loadModuleOnServer } from "./players/all.js";
export type * from "./Player.js";
