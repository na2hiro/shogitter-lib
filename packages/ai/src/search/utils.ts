import { ShogitterAiException } from "../ShogitterAiException.js";

export function randomSelect<T>(arr: T[]) {
  if (arr.length === 0)
    throw new ShogitterAiException("randomSelect: empty array");
  return arr[Math.floor(Math.random() * arr.length)];
}
