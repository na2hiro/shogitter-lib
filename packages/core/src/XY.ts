import { ShogitterCoreException } from "./utils/phpCompat.js";
import { num2kan_decimal } from "./MyLib.js";
import { Direction } from "./Direction.js";

export enum RelativeType {
  TATEYOKO = 1,
  ALLROUND = 2,
  HALFROUND = 3,
}

export const XYMatcher = (x: number, y: number) => (xy: XY) =>
  xy.x == x && xy.y == y;

export type XYObj = [number, number];

export class RelXY {
  x: number;
  y: number;
  static constants = {
    [RelativeType.TATEYOKO]: [
      //縦横１マスの位置（計４マス）
      new RelXY(1, 0),
      new RelXY(-1, 0),
      new RelXY(0, 1),
      new RelXY(0, -1),
    ],
    [RelativeType.ALLROUND]: [
      //周囲１マスの位置（計８マス）
      new RelXY(1, 0),
      new RelXY(-1, 0),
      new RelXY(0, 1),
      new RelXY(0, -1),
      new RelXY(1, 1),
      new RelXY(1, -1),
      new RelXY(-1, 1),
      new RelXY(-1, -1),
    ],
    [RelativeType.HALFROUND]: [
      //周囲１マスの位置を半周（計４マス）
      new RelXY(1, 0),
      new RelXY(1, 1),
      new RelXY(0, 1),
      new RelXY(-1, 1),
    ],
  };
  static getVector(type: RelativeType): RelXY[] {
    return RelXY.constants[type];
  }

  static fromArray([x, y]: XYObj) {
    return new RelXY(x, y);
  }

  constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }

  toString() {
    return `(${this.x}, ${this.y}) `;
  }

  getHash() {
    return `${this.x},${this.y}`;
  }

  add(x: number, y: number) {
    this.x += x;
    this.y += y;
  }

  addRel(relXY: RelXY, tani = 1) {
    this.x += relXY.x * tani;
    this.y += relXY.y * tani;
  }

  set(x: number, y: number) {
    this.x = x;
    this.y = y;
  }

  equals(xy: XY) {
    return this.x == xy.x && this.y == xy.y;
  }

  getClone(x = 0, y = 0) {
    const tmp = new RelXY(this.x, this.y);
    tmp.add(x, y);
    return tmp;
  }

  getCloneRel(relXY: RelXY, tani = 1) {
    const tmp = new RelXY(this.x, this.y);
    tmp.addRel(relXY, tani);
    return tmp;
  }

  getArray(): XYObj {
    return [this.x, this.y];
  }

  turn(direction: Direction) {
    let x, y;
    switch (direction) {
      case Direction.BLACK: //先手0度回転
        x = this.x;
        y = this.y;
        break;
      case Direction.WHITE: //後手180度回転
        x = -this.x;
        y = -this.y;
        break;
      case Direction.LEFT: //左手90度回転
        x = this.y;
        y = -this.x;
        break;
      case Direction.RIGHT: //右手270度回転
        x = -this.y;
        y = this.x;
        break;
      default:
        throw new ShogitterCoreException(
          `駒の向き${direction}が定義されていません。`
        );
    }
    this.x = x;
    this.y = y;
  }

  turnReverse(direction: Direction) {
    let x, y;
    switch (direction) {
      case Direction.BLACK: //先手0度回転
        x = this.x;
        y = this.y;
        break;
      case Direction.WHITE: //後手180度回転
        x = -this.x;
        y = -this.y;
        break;
      case Direction.LEFT: //左手270度回転
        x = -this.y;
        y = this.x;
        break;
      case Direction.RIGHT: //右手90度回転
        x = this.y;
        y = -this.x;
        break;
      default:
        throw new ShogitterCoreException(
          `駒の向き${direction}が定義されていません。`,
          1
        );
    }
    this.x = x;
    this.y = y;
  }

  //単位ベクトル化(最小公倍数で割る)
  unify() {
    let abs;
    if (this.x == 0) {
      if (this.y == 0) {
        return 1;
      } else {
        abs = Math.abs(this.y);
        this.y /= abs;
        return abs;
      }
    } else {
      if (this.y == 0) {
        abs = Math.abs(this.x);
        this.x /= abs;
        return abs;
      } else {
        const gcd = RelXY.gcd(Math.abs(this.x), Math.abs(this.y));
        this.x /= gcd;
        this.y /= gcd;
        return gcd;
      }
    }
  }

  //最大公倍数を求める（ユークリッドの互除法）
  static gcd(m: number, n: number) {
    while (m % n != 0) {
      const temp = n;
      n = m % n;
      m = temp;
    }
    return n;
  }

  getCloneFlipped() {
    const ret = this.getClone();
    ret.x *= -1;
    ret.y *= -1;
    return ret;
  }
}

export default class XY extends RelXY {
  /**
   * 座標を生成
   * @param x
   * @param y
   */
  constructor(x: number, y: number) {
    super(x, y);
  }

  static fromArray([x, y]: XYObj) {
    return new XY(x, y);
  }

  toArray(): XYObj {
    return [this.x, this.y];
  }

  toString() {
    return `(${this.x}, ${this.y})`;
  }

  getFormat() {
    return this.x + num2kan_decimal(`${this.y}`);
  }

  turn() {
    throw new ShogitterCoreException("座標が回転できません。");
  }

  turnReverse() {
    throw new ShogitterCoreException("座標が回転できません。");
  }

  getDistance(xy: XY) {
    return Math.pow(this.x - xy.x, 2) + Math.pow(this.y - xy.y, 2);
  }

  to(xy: XY) {
    return new RelXY(xy.x - this.x, xy.y - this.y);
  }

  from(xy: XY) {
    return new RelXY(this.x - xy.x, this.y - xy.y);
  }

  clone() {
    return new XY(this.x, this.y);
  }

  getClone(x = 0, y = 0): XY {
    const tmp = new XY(this.x, this.y);
    tmp.add(x, y);
    return tmp;
  }

  getCloneRel(relXY: RelXY, tani = 1) {
    const tmp = new XY(this.x, this.y);
    tmp.addRel(relXY, tani);
    return tmp;
  }
}
