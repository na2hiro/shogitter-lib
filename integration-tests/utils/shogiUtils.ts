import Shogi from "../../src/Shogi";
import XY from "../../src/XY";
import {Species} from "../../src/Ban";
import {serialize} from "./ShogiSerializer";
import {Direction} from "../../src/Direction";

export function move(shogi: Shogi, fromX: number, fromY: number, toX: number, toY: number, nari = false) {
    try {
        rawMove(shogi, fromX, fromY, toX, toY, nari);
    } catch (e) {
        console.error(serialize(shogi, null, null, null, null, null))
        throw e;
    }
}

export function rawMove(shogi: Shogi, fromX: number, fromY: number, toX: number, toY: number, nari = false) {
    shogi.runCommand({
        type: "move",
        from: [fromX, fromY],
        to: [toX, toY],
        nari
    })
}

export function put(shogi: Shogi, toX: number, toY: number, species: Species, id?: number) {
    try {
        rawPut(shogi, toX, toY, species, id);
    } catch (e) {
        console.error(serialize(shogi, null, null, null, null, null))
        throw e;
    }
}

export function rawPut(shogi: Shogi, toX: number, toY: number, species: Species, id?: number) {
    shogi.runCommand({
        type: "put",
        to: [toX, toY],
        put: species,
        direction: shogi.teban.getNowDirection(),
        id
    })
}

export function assertCellIsNull(shogi: Shogi, x: number, y: number) {
    expect(shogi.ban.get(new XY(x, y)).isNull()).toBeTruthy;
}

export function assertCell(shogi: Shogi, x: number, y: number, direction: Direction, species: Species) {
    const koma = shogi.ban.get(new XY(x, y));
    expect(koma.isNull()).toBeFalsy();
    expect(koma.species).toBe(species);
    expect(koma.direction).toBe(direction);
}
