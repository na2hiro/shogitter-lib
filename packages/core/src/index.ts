import Shogi from "./Shogi.js";
import { Species } from "./Ban.js";
import { PlayerInfo } from "./Teban.js";
import { shogitterDB } from "./ShogitterDB.js";
import XY from "./XY.js";
import { KomaInfo } from "./db/koma.js";
import * as quantumUtils from "./utils/quantumUtils.js";

export * from "./Shogi.js";
export * from "./Command.js";
export * from "./Direction.js";
export * from "./Koma.js";

export { Shogi, Species, KomaInfo, PlayerInfo, shogitterDB, XY, quantumUtils };
