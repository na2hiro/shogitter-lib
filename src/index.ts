import Shogi from "./Shogi";
import {Direction} from "./Direction";
import {Species} from "./Ban";
import { PlayerInfo } from "./Teban";
import { KomaInfo, shogitterDB } from "./ShogitterDB";
import XY from "./XY";

export * from "./Shogi";
export default Shogi;
export {Direction, Species, KomaInfo, PlayerInfo, shogitterDB, XY};