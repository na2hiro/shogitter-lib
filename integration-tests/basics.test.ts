import Shogi from "../src/Shogi";
import XY, { XYMatcher } from "../src/XY";
import { Species } from "../src/Ban";
import { move, rawMove, put } from "./utils/shogiUtils";
import { beforeEach } from "vitest";
import { PlayerInfo } from "../src";

describe("Basic", () => {
  it("can move pieces", () => {
    const shogi = Shogi.ofRuleId(0);
    expect(shogi).toMatchSnapshot();
    shogi.start();

    moveAndSnapshot(shogi, 7, 7, 7, 6);
    moveAndSnapshot(shogi, 3, 3, 3, 4);
    moveAndSnapshot(shogi, 8, 8, 2, 2, true);
    moveAndSnapshot(shogi, 3, 1, 2, 2);
    putAndSnapshot(shogi, 4, 5, "af");
    moveAndSnapshot(shogi, 4, 1, 3, 2);
    moveAndSnapshot(shogi, 4, 5, 3, 4);
    putAndSnapshot(shogi, 9, 5, "af");
    moveAndSnapshot(shogi, 7, 9, 6, 8);
  });
  it("throws Error for invalid move", () => {
    const shogi = Shogi.ofRuleId(0);
    shogi.start();

    expect(() => {
      rawMove(shogi, 7, 7, 7, 5);
    }).toThrow();
  });

  it("can judge tsumi", () => {
    const shogi = Shogi.ofRuleId(0);
    expect(shogi).toMatchSnapshot();
    shogi.start();

    move(shogi, 7, 7, 7, 6);
    move(shogi, 5, 3, 5, 4);
    move(shogi, 8, 8, 3, 3, false);
    move(shogi, 3, 1, 4, 2);
    move(shogi, 3, 3, 4, 2, false);
    move(shogi, 5, 1, 5, 2);
    putAndSnapshot(shogi, 5, 3, "ad");
  });

  it("is isomorphic on serialization/deserialization", () => {
    let shogi = Shogi.ofRuleId(0);
    shogi.start();
    shogi = isIso(shogi);
    move(shogi, 7, 7, 7, 6);
    shogi = isIso(shogi);
    move(shogi, 3, 3, 3, 4);
    shogi = isIso(shogi);
    move(shogi, 8, 8, 2, 2, true);
    shogi = isIso(shogi);
    move(shogi, 3, 1, 2, 2);
    shogi = isIso(shogi);
    put(shogi, 4, 5, "af");
    shogi = isIso(shogi);
    move(shogi, 4, 1, 3, 2);
    shogi = isIso(shogi);

    function isIso(shogi: Shogi) {
      const obj = shogi.getObject();
      const shogi2 = Shogi.ofJkf(obj);
      const obj2 = shogi2.getObject();
      expect(obj).toEqual(obj2);
      return shogi2;
    }
  });

  describe("player info", function () {
    const playerWithUser1: PlayerInfo = {
      user: [{ name: "na2hiro", id: 12345 }],
    };
    const playerWithUser2: PlayerInfo = {
      user: [{ name: "SHOSTAKOVI_CH", id: 67890 }],
    };
    const playerWithoutUser: PlayerInfo = { user: [null] };
    describe.each([
      {
        name: "when both players sit",
        playerInfo: [playerWithUser1, playerWithUser2],
      },
      {
        name: "when one of the players sits",
        playerInfo: [playerWithUser1, playerWithoutUser],
      },
    ])("$name", ({ playerInfo }) => {
      let shogi: Shogi;
      beforeEach(() => {
        shogi = Shogi.ofRuleId(0);
        shogi.teban.setArrayPlayerInfo(JSON.parse(JSON.stringify(playerInfo)));
      });

      it("can be read", () => {
        expect(shogi.teban.getArrayPlayerInfo()).toEqual(playerInfo);
      });

      it("is kept after reset", () => {
        shogi.constructById(1);
        const actual = shogi.teban.getArrayPlayerInfo();
        expect(actual).toEqual(playerInfo);
      });

      it("is kept after changing turn", () => {
        shogi.teban.changeDirection();
        const actual = shogi.teban.getArrayPlayerInfo();
        expect(actual).toEqual([playerInfo[1], playerInfo[0]]);
      });
    });
  });

  function moveAndSnapshot(
    shogi: Shogi,
    fromX: number,
    fromY: number,
    toX: number,
    toY: number,
    nari = false
  ) {
    const from = new XY(fromX, fromY);
    const movables = shogi.ban.get(from).getMovable();
    expect(movables).toMatchSnapshot();
    const to = movables.filter((move) => XYMatcher(toX, toY)(move.XY))[0].XY;
    expect(to).toBeTruthy();
    shogi.move(from, to, nari);
    expect(shogi).toMatchSnapshot();
  }

  function putAndSnapshot(
    shogi: Shogi,
    toX: number,
    toY: number,
    species: Species
  ) {
    shogi.put(new XY(toX, toY), species, shogi.teban.get());
    expect(shogi).toMatchSnapshot();
  }
});
