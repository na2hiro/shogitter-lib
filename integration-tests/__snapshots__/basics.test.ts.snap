// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic can judge tsumi 1`] = `
ban:
(1) 
1ab1ac1ad1ae1ah1ae1ad1ac1ab
___1ag_______________1af___
1aa1aa1aa1aa1aa1aa1aa1aa1aa
___________________________
___________________________
___________________________
0aa0aa0aa0aa0aa0aa0aa0aa0aa
___0af_______________0ag___
0ab0ac0ad0ae0ah0ae0ad0ac0ab
(0) 
kifu:

{
  "version": "0.0",
  "status": {
    "num": 0
  },
  "ruleid": 0,
  "teban": 0,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Basic can judge tsumi 2`] = `
ban:
(1) 
1ab1ac1ad1ae___1ae___1ac1ab
___1ag______1ah0af___1af___
1aa1aa1aa1aa0ad1aa___1aa1aa
____________1aa____________
___________________________
______0aa__________________
0aa0aa___0aa0aa0aa0aa0aa0aa
_____________________0ag___
0ab0ac0ad0ae0ah0ae0ad0ac0ab
(0) aa1 
kifu:
☗7六歩 [0,[7,6,[],[0,"aa"]],[7,7,[0,"aa"],[]]]
☖5四歩 [1,[5,4,[],[1,"aa"]],[5,3,[1,"aa"],[]]]
☗3三角不成 [0,[3,3,[1,"aa"],[0,"af"]],[8,8,[0,"af"],[]],[1,0,"aa"]]
☖4二銀 [1,[4,2,[],[1,"ad"]],[3,1,[1,"ad"],[]]]
☗同角不成 [0,[4,2,[1,"ad"],[0,"af"]],[3,3,[0,"af"],[]],[1,0,"ad"]]
☖5二玉 [1,[5,2,[],[1,"ah"]],[5,1,[1,"ah"],[]]]
☗5三銀 [0,[5,3,[],[0,"ad"]],[-1,0,"ad"]]
☖詰み "_1"
{
  "version": "0.0",
  "status": {
    "num": 2,
    "message": "後手玉が詰みました。先手の勝ちです。"
  },
  "ruleid": 0,
  "teban": 1,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Basic can move pieces 1`] = `
ban:
(1) 
1ab1ac1ad1ae1ah1ae1ad1ac1ab
___1ag_______________1af___
1aa1aa1aa1aa1aa1aa1aa1aa1aa
___________________________
___________________________
___________________________
0aa0aa0aa0aa0aa0aa0aa0aa0aa
___0af_______________0ag___
0ab0ac0ad0ae0ah0ae0ad0ac0ab
(0) 
kifu:

{
  "version": "0.0",
  "status": {
    "num": 0
  },
  "ruleid": 0,
  "teban": 0,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Basic can move pieces 2`] = `
Array [
  Object {
    "XY": (7, 6),
    "type": 1,
  },
]
`;

exports[`Basic can move pieces 3`] = `
ban:
(1) 
1ab1ac1ad1ae1ah1ae1ad1ac1ab
___1ag_______________1af___
1aa1aa1aa1aa1aa1aa1aa1aa1aa
___________________________
___________________________
______0aa__________________
0aa0aa___0aa0aa0aa0aa0aa0aa
___0af_______________0ag___
0ab0ac0ad0ae0ah0ae0ad0ac0ab
(0) 
kifu:
☗7六歩 [0,[7,6,[],[0,"aa"]],[7,7,[0,"aa"],[]]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 0,
  "teban": 1,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Basic can move pieces 4`] = `
Array [
  Object {
    "XY": (3, 4),
    "type": 1,
  },
]
`;

exports[`Basic can move pieces 5`] = `
ban:
(1) 
1ab1ac1ad1ae1ah1ae1ad1ac1ab
___1ag_______________1af___
1aa1aa1aa1aa1aa1aa___1aa1aa
__________________1aa______
___________________________
______0aa__________________
0aa0aa___0aa0aa0aa0aa0aa0aa
___0af_______________0ag___
0ab0ac0ad0ae0ah0ae0ad0ac0ab
(0) 
kifu:
☗7六歩 [0,[7,6,[],[0,"aa"]],[7,7,[0,"aa"],[]]]
☖3四歩 [1,[3,4,[],[1,"aa"]],[3,3,[1,"aa"],[]]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 0,
  "teban": 0,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Basic can move pieces 6`] = `
Array [
  Object {
    "XY": (9, 7),
    "type": 2,
  },
  Object {
    "XY": (7, 7),
    "type": 2,
  },
  Object {
    "XY": (6, 6),
    "type": 2,
  },
  Object {
    "XY": (5, 5),
    "type": 2,
  },
  Object {
    "XY": (4, 4),
    "type": 2,
  },
  Object {
    "XY": (3, 3),
    "type": 2,
  },
  Object {
    "XY": (2, 2),
    "type": 2,
  },
  Object {
    "XY": (9, 9),
    "type": 2,
  },
  Object {
    "XY": (7, 9),
    "type": 2,
  },
]
`;

exports[`Basic can move pieces 7`] = `
ban:
(1) 
1ab1ac1ad1ae1ah1ae1ad1ac1ab
___1ag_______________0am___
1aa1aa1aa1aa1aa1aa___1aa1aa
__________________1aa______
___________________________
______0aa__________________
0aa0aa___0aa0aa0aa0aa0aa0aa
_____________________0ag___
0ab0ac0ad0ae0ah0ae0ad0ac0ab
(0) af1 
kifu:
☗7六歩 [0,[7,6,[],[0,"aa"]],[7,7,[0,"aa"],[]]]
☖3四歩 [1,[3,4,[],[1,"aa"]],[3,3,[1,"aa"],[]]]
☗2二角成 [0,[2,2,[1,"af"],[0,"am"]],[8,8,[0,"af"],[]],[1,0,"af"]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 0,
  "teban": 1,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Basic can move pieces 8`] = `
Array [
  Object {
    "XY": (2, 2),
    "type": 1,
  },
  Object {
    "XY": (3, 2),
    "type": 1,
  },
  Object {
    "XY": (4, 2),
    "type": 1,
  },
]
`;

exports[`Basic can move pieces 9`] = `
ban:
(1) af1 
1ab1ac1ad1ae1ah1ae___1ac1ab
___1ag_______________1ad___
1aa1aa1aa1aa1aa1aa___1aa1aa
__________________1aa______
___________________________
______0aa__________________
0aa0aa___0aa0aa0aa0aa0aa0aa
_____________________0ag___
0ab0ac0ad0ae0ah0ae0ad0ac0ab
(0) af1 
kifu:
☗7六歩 [0,[7,6,[],[0,"aa"]],[7,7,[0,"aa"],[]]]
☖3四歩 [1,[3,4,[],[1,"aa"]],[3,3,[1,"aa"],[]]]
☗2二角成 [0,[2,2,[1,"af"],[0,"am"]],[8,8,[0,"af"],[]],[1,0,"af"]]
☖同銀 [1,[2,2,[0,"am"],[1,"ad"]],[3,1,[1,"ad"],[]],[1,1,"af"]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 0,
  "teban": 0,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Basic can move pieces 10`] = `
ban:
(1) af1 
1ab1ac1ad1ae1ah1ae___1ac1ab
___1ag_______________1ad___
1aa1aa1aa1aa1aa1aa___1aa1aa
__________________1aa______
_______________0af_________
______0aa__________________
0aa0aa___0aa0aa0aa0aa0aa0aa
_____________________0ag___
0ab0ac0ad0ae0ah0ae0ad0ac0ab
(0) 
kifu:
☗7六歩 [0,[7,6,[],[0,"aa"]],[7,7,[0,"aa"],[]]]
☖3四歩 [1,[3,4,[],[1,"aa"]],[3,3,[1,"aa"],[]]]
☗2二角成 [0,[2,2,[1,"af"],[0,"am"]],[8,8,[0,"af"],[]],[1,0,"af"]]
☖同銀 [1,[2,2,[0,"am"],[1,"ad"]],[3,1,[1,"ad"],[]],[1,1,"af"]]
☗4五角 [0,[4,5,[],[0,"af"]],[-1,0,"af"]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 0,
  "teban": 1,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Basic can move pieces 11`] = `
Array [
  Object {
    "XY": (3, 2),
    "type": 1,
  },
  Object {
    "XY": (4, 2),
    "type": 1,
  },
  Object {
    "XY": (5, 2),
    "type": 1,
  },
  Object {
    "XY": (3, 1),
    "type": 1,
  },
  Object {
    "XY": (5, 1),
    "type": 1,
  },
]
`;

exports[`Basic can move pieces 12`] = `
ban:
(1) af1 
1ab1ac1ad1ae1ah______1ac1ab
___1ag____________1ae1ad___
1aa1aa1aa1aa1aa1aa___1aa1aa
__________________1aa______
_______________0af_________
______0aa__________________
0aa0aa___0aa0aa0aa0aa0aa0aa
_____________________0ag___
0ab0ac0ad0ae0ah0ae0ad0ac0ab
(0) 
kifu:
☗7六歩 [0,[7,6,[],[0,"aa"]],[7,7,[0,"aa"],[]]]
☖3四歩 [1,[3,4,[],[1,"aa"]],[3,3,[1,"aa"],[]]]
☗2二角成 [0,[2,2,[1,"af"],[0,"am"]],[8,8,[0,"af"],[]],[1,0,"af"]]
☖同銀 [1,[2,2,[0,"am"],[1,"ad"]],[3,1,[1,"ad"],[]],[1,1,"af"]]
☗4五角 [0,[4,5,[],[0,"af"]],[-1,0,"af"]]
☖3二金 [1,[3,2,[],[1,"ae"]],[4,1,[1,"ae"],[]]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 0,
  "teban": 0,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Basic can move pieces 13`] = `
Array [
  Object {
    "XY": (5, 4),
    "type": 2,
  },
  Object {
    "XY": (6, 3),
    "type": 2,
  },
  Object {
    "XY": (3, 4),
    "type": 2,
  },
  Object {
    "XY": (5, 6),
    "type": 2,
  },
  Object {
    "XY": (6, 7),
    "type": 2,
  },
  Object {
    "XY": (3, 6),
    "type": 2,
  },
  Object {
    "XY": (2, 7),
    "type": 2,
  },
]
`;

exports[`Basic can move pieces 14`] = `
ban:
(1) af1 
1ab1ac1ad1ae1ah______1ac1ab
___1ag____________1ae1ad___
1aa1aa1aa1aa1aa1aa___1aa1aa
__________________0af______
___________________________
______0aa__________________
0aa0aa___0aa0aa0aa0aa0aa0aa
_____________________0ag___
0ab0ac0ad0ae0ah0ae0ad0ac0ab
(0) aa1 
kifu:
☗7六歩 [0,[7,6,[],[0,"aa"]],[7,7,[0,"aa"],[]]]
☖3四歩 [1,[3,4,[],[1,"aa"]],[3,3,[1,"aa"],[]]]
☗2二角成 [0,[2,2,[1,"af"],[0,"am"]],[8,8,[0,"af"],[]],[1,0,"af"]]
☖同銀 [1,[2,2,[0,"am"],[1,"ad"]],[3,1,[1,"ad"],[]],[1,1,"af"]]
☗4五角 [0,[4,5,[],[0,"af"]],[-1,0,"af"]]
☖3二金 [1,[3,2,[],[1,"ae"]],[4,1,[1,"ae"],[]]]
☗3四角 [0,[3,4,[1,"aa"],[0,"af"]],[4,5,[0,"af"],[]],[1,0,"aa"]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 0,
  "teban": 1,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Basic can move pieces 15`] = `
ban:
(1) 
1ab1ac1ad1ae1ah______1ac1ab
___1ag____________1ae1ad___
1aa1aa1aa1aa1aa1aa___1aa1aa
__________________0af______
1af________________________
______0aa__________________
0aa0aa___0aa0aa0aa0aa0aa0aa
_____________________0ag___
0ab0ac0ad0ae0ah0ae0ad0ac0ab
(0) aa1 
kifu:
☗7六歩 [0,[7,6,[],[0,"aa"]],[7,7,[0,"aa"],[]]]
☖3四歩 [1,[3,4,[],[1,"aa"]],[3,3,[1,"aa"],[]]]
☗2二角成 [0,[2,2,[1,"af"],[0,"am"]],[8,8,[0,"af"],[]],[1,0,"af"]]
☖同銀 [1,[2,2,[0,"am"],[1,"ad"]],[3,1,[1,"ad"],[]],[1,1,"af"]]
☗4五角 [0,[4,5,[],[0,"af"]],[-1,0,"af"]]
☖3二金 [1,[3,2,[],[1,"ae"]],[4,1,[1,"ae"],[]]]
☗3四角 [0,[3,4,[1,"aa"],[0,"af"]],[4,5,[0,"af"],[]],[1,0,"aa"]]
☖9五角 [1,[9,5,[],[1,"af"]],[-1,1,"af"]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 0,
  "teban": 0,
  "turn": 0,
  "date": {},
  "moving": null
}
`;

exports[`Basic can move pieces 16`] = `
Array [
  Object {
    "XY": (8, 8),
    "type": 1,
  },
  Object {
    "XY": (7, 8),
    "type": 1,
  },
  Object {
    "XY": (6, 8),
    "type": 1,
  },
]
`;

exports[`Basic can move pieces 17`] = `
ban:
(1) 
1ab1ac1ad1ae1ah______1ac1ab
___1ag____________1ae1ad___
1aa1aa1aa1aa1aa1aa___1aa1aa
__________________0af______
1af________________________
______0aa__________________
0aa0aa___0aa0aa0aa0aa0aa0aa
_________0ad_________0ag___
0ab0ac___0ae0ah0ae0ad0ac0ab
(0) aa1 
kifu:
☗7六歩 [0,[7,6,[],[0,"aa"]],[7,7,[0,"aa"],[]]]
☖3四歩 [1,[3,4,[],[1,"aa"]],[3,3,[1,"aa"],[]]]
☗2二角成 [0,[2,2,[1,"af"],[0,"am"]],[8,8,[0,"af"],[]],[1,0,"af"]]
☖同銀 [1,[2,2,[0,"am"],[1,"ad"]],[3,1,[1,"ad"],[]],[1,1,"af"]]
☗4五角 [0,[4,5,[],[0,"af"]],[-1,0,"af"]]
☖3二金 [1,[3,2,[],[1,"ae"]],[4,1,[1,"ae"],[]]]
☗3四角 [0,[3,4,[1,"aa"],[0,"af"]],[4,5,[0,"af"],[]],[1,0,"aa"]]
☖9五角 [1,[9,5,[],[1,"af"]],[-1,1,"af"]]
☗6八銀 [0,[6,8,[],[0,"ad"]],[7,9,[0,"ad"],[]]]
{
  "version": "0.0",
  "status": {
    "num": 1
  },
  "ruleid": 0,
  "teban": 1,
  "turn": 0,
  "date": {},
  "moving": null
}
`;
